apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/FetchPolicyInterceptors.kt

```diff
@@ -186,28 +187,29 @@ val CacheAndNetworkInterceptor = object : ApolloInterceptor {
             } else {
               throw it
             }
-          }.singleOrNull()
-
-      if (networkResponse != null) {
-        emit(
-            networkResponse.newBuilder()
+          }
+          .map { response ->
+            response.newBuilder()
                 .cacheInfo(
-                    networkResponse.cacheInfo!!
+                    response.cacheInfo!!
                         .newBuilder()
                         .cacheMissException(cacheException as? CacheMissException)
                         .build()
                 )
                 .build()
-        )
-        return@flow
-      }
-      if (cacheException != null) {
-        throw ApolloCompositeException(
-            first = cacheException,
-            second = networkException
-        )
+          }
+
+      emitAll(networkResponse)
+
+      if (networkException != null) {
+        if (cacheException != null) {
+          throw ApolloCompositeException(
+              first = cacheException,
+              second = networkException
+          )
+        }
+        throw networkException!!
       }
-      throw networkException!!
```

Comment:
Not a fan of what all the custom logic in this file has become. In retrospect, I think that throwing on cache misses was a mistake (see https://github.com/apollographql/apollo-kotlin/issues/3976) but throwing on network errors might also be (because throwing always break the flow).

If we emit `ApolloResponse` all the time (with a potential `.exception` field) we can move the `ApolloCompositeException` logic to `.execute()`:

```kotlin
  suspend fun execute(): ApolloResponse<D> {
    val responses = toFlow().toList()
    
    val (errors, successes) = responses.partition { it.exception != null }
    
    if (successes.size > 1) {
      throw ApolloException("The operation returned multiple items, use .toFlow() instead of .execute()")
    } else if (successes.isEmpty()) {
      if (errors.size == 1) {
        throw errors.single().exception!!
      } else if (errors.size > 1) {
        throw ApolloCompositeException(errors.map { it.exception })
      } else {
        throw ApolloException("The operation did not emit any item, check your interceptor chain")
      }
    } else {
      return successes.first()
    }
  }
```

This would put this logic in a central and final place and keep the flow running for as long as possible.

Reply:
I played with the idea [here](https://github.com/apollographql/apollo-kotlin/commit/587167a16e81e04d6a2261613ba6beb8a229a984), and this indeed simplifies `FetchPolicyInterceptors` + it's nice to have the exception handling centralized here.

This breaks some tests though and I'm not sure we can do this change without changing the contract of some of the calls due to how they currently throw exceptions in `toFlow` vs the idea of emitting them with a non-null `exception` field instead.

Ok for me to open an issue to keep track of this idea?

Reply:
Sure thing! It's a big change that might require some tweaks (maybe even breaking changes). Let's follow up in a separate issue

Reply:
I've opened https://github.com/apollographql/apollo-kotlin/issues/4003.